src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -500,10 +500,14 @@ private async Task ProcessIncomingFramesAsync()
 
                     // Process the initial SETTINGS frame. This will send an ACK.
                     ProcessSettingsFrame(frameHeader, initialFrame: true);
+
+                    Debug.Assert(InitialSettingsReceived.Task.IsCompleted);
                 }
-                catch (IOException e)
+                catch (Exception e)
                 {
-                    throw new IOException(SR.net_http_http2_connection_not_established, e);
+                    e = new IOException(SR.net_http_http2_connection_not_established, e);
+                    InitialSettingsReceived.TrySetException(e);
+                    throw e;
```

Comment:
The same exception instance shouldn't be used twice, they're not thread safe. The `throw` and the `TrySetException` need to be passed different instances.

Reply:
Which part do you mean is not thread-safe? We share exceptions like this all the time.

Reply:
Given the right (wrong) interleaving of concurrent use with ExceptionDispathInfo.Capture/Throws (as is used by async/await under the covers), you can end up with call stacks that have sections from different logical flows of execution and in orders that don't make sense.  We've generally accepted that.

Reply:
> you can end up with call stacks that have sections from different logical flows of execution

Does that apply in this case given that `TrySetException` will do the capture before we get to the throw?

Reply:
We had a similar issue in Kestrel, passing the same exception instance to multiple TCS's. We ended up cloning or wrapping the exceptions so each TCS got a unique instance.

Reply:
My understanding is that in cases like this where we do it in series (`TrySetException` first, then throw) it shouldn't be an issue as the `TrySetException` will do the capture before we get to the throw. Is that not the case?

We have some other cases where we'll share the same exception like [here](https://github.com/dotnet/runtime/blob/ffb3d8d49c4d3d71ec4a0a7b595e0d1a233a0a74/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1077-L1080), where that exception may be thrown from multiple threads at the same time. I take it that would be more problematic?

Reply:
And that is only a problem for the top-most exception?
That is, an inner exception can be safely shared as long as the outer wrapper instance is different?

Reply:
An inner exception is just arbitrary additional data reference by a wrapping Exception; it's not read or written in any way as part of throws/catches for another exception that happens to be storing a reference to it.

Reply:
Okay, thanks for the clarifications! I'll update the code.